File: app/src/main/java/kr/asv/loancalculator/LoanCalculator.java
package kr.asv.loancalculator;

import java.math.BigInteger;

/**
 * 대출이자 계산기
 */
public class LoanCalculator
{
    /**
     * Options 값들. 가족수, 자녀수, 비과세액 등
     */
    private final LoanCalculatorOptions options = new LoanCalculatorOptions();

    @SuppressWarnings("unused")
    public enum AmortizationMethods {
        FULL_AMORTIZATION, EQUAL_PRINCIPAL
    }

    private PaymentSchedules schedules;

    private BigInteger summaryInterest;
    /**
     * 생성자
     */
    public LoanCalculator()
    {
        //this.initialize();
    }

    /**
     * 계산 실행
     */
    public void run()
    {
        if (options.isDebug())
        {
            debug(options);
        }
        schedules = new PaymentSchedules();
        Amortization amortization;
        if (options.getAmortizationMethod() == AmortizationMethods.EQUAL_PRINCIPAL)
        {
            amortization = new EqualPrincipalAmortization();
            //schedules = equalPrincipalAmortization.calculate(options);// 원금 균등 분할 상환
        }
        else
        {
            amortization = new FullAmortization();
            //schedules = fullAmortization.calculate(options);// 원리금 균등 분할 상환
        }
        amortization.calculate(options);
        schedules = amortization.getSchedules();
        summaryInterest = amortization.getSummaryInterest();
    }

    private void debug(Object obj)
    {
        System.out.println(obj);
    }

    public LoanCalculatorOptions getOptions()
    {
        return this.options;
    }

    @SuppressWarnings("UnusedReturnValue")
    public PaymentSchedules getSchedules()
    {
        return schedules;
    }

    @SuppressWarnings("UnusedReturnValue")
    public BigInteger getSummaryInterest()
    {
        return summaryInterest;
    }

}

File: app/src/main/java/kr/asv/loancalculator/LoanCalculator.kt
package kr.asv.loancalculator

import java.math.BigInteger

/**
 * 대출이자 계산기 클래스
 */
class LoanCalculator {
    /**
     * Options 값들. 가족수, 자녀수, 비과세액 등
     */
    val options = LoanCalculatorOptions()

    enum class AmortizationMethods {
        FULL_AMORTIZATION, EQUAL_PRINCIPAL
    }

    var schedules: PaymentSchedules? = null
        private set

    var summaryInterest: BigInteger? = null
        private set

    /**
     * 계산 실행
     */
    fun run() {
        if (options.isDebug) {
            debug(options)
        }
        //schedules = PaymentSchedules()
        schedules?.clear() // 계산 전에 남아있는 것이 있다면 clear


        val amortization: Amortization = if (options.amortizationMethod == AmortizationMethods.EQUAL_PRINCIPAL) {
            EqualPrincipalAmortization()
            //schedules = equalPrincipalAmortization.calculate(options);// 원금 균등 분할 상환
        } else {
            FullAmortization()
            //schedules = fullAmortization.calculate(options);// 원리금 균등 분할 상환
        }
        amortization.calculate(options)

        // 결과
        schedules = amortization.schedules
        summaryInterest = amortization.summaryInterest
    }

    private fun debug(obj: Any) {
        println(obj)
    }
}